api/v1beta2/receiver_types.go

```diff
@@ -0,0 +1,143 @@
+	ReceiverWebhookPath string = "/hook/"
+	GenericReceiver     string = "generic"
+	GenericHMACReceiver string = "generic-hmac"
+	GitHubReceiver      string = "github"
+	GitLabReceiver      string = "gitlab"
+	BitbucketReceiver   string = "bitbucket"
+	HarborReceiver      string = "harbor"
+	DockerHubReceiver   string = "dockerhub"
+	QuayReceiver        string = "quay"
+	GCRReceiver         string = "gcr"
+	NexusReceiver       string = "nexus"
+	ACRReceiver         string = "acr"
+)
+
+// ReceiverSpec defines the desired state of the Receiver.
+type ReceiverSpec struct {
+	// Type of webhook sender, used to determine
+	// the validation procedure and payload deserialization.
+	// +kubebuilder:validation:Enum=generic;generic-hmac;github;gitlab;bitbucket;harbor;dockerhub;quay;gcr;nexus;acr
+	// +required
+	Type string `json:"type"`
+
+	// Interval at which to reconcile the Receiver with its Secret references.
+	// +kubebuilder:default="10m"
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ms|s|m|h))+$"
+	// +optional
+	Interval metav1.Duration `json:"interval"`
+
+	// Events specifies the list of event types to handle,
+	// e.g. 'push' for GitHub or 'Push Hook' for GitLab.
+	// +optional
+	Events []string `json:"events"`
+
+	// A list of resources to be notified about changes.
+	// +required
+	Resources []CrossNamespaceObjectReference `json:"resources"`
+
+	// SecretRef specifies the Secret containing the token used
+	// to validate the payload authenticity.
+	// +required
+	SecretRef meta.LocalObjectReference `json:"secretRef,omitempty"`
+
+	// Suspend tells the controller to suspend subsequent
+	// events handling for this receiver.
+	// +optional
+	Suspend bool `json:"suspend,omitempty"`
+}
+
+// ReceiverStatus defines the observed state of the Receiver.
+type ReceiverStatus struct {
+	meta.ReconcileRequestStatus `json:",inline"`
+
+	// Conditions holds the conditions for the Receiver.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// URL is the generated incoming webhook address in the format
+	// of '/hook/sha256sum(token+name+namespace)'.
+	// +optional
+	URL string `json:"url,omitempty"`
+
+	// ObservedGeneration is the last observed generation of the Receiver object.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+// GetConditions returns the status conditions of the object.
+func (in *Receiver) GetConditions() []metav1.Condition {
+	return in.Status.Conditions
+}
+
+// SetConditions sets the status conditions on the object.
+func (in *Receiver) SetConditions(conditions []metav1.Condition) {
+	in.Status.Conditions = conditions
+}
+
+// GetWebhookURL returns the incoming webhook URL for the given token.
+func (in *Receiver) GetWebhookURL(token string) string {
+	digest := sha256.Sum256([]byte(token + in.GetName() + in.GetNamespace()))
+	return fmt.Sprintf("%s%x", ReceiverWebhookPath, digest)
```

Comment:
This is a new API method and it returns the webhook path, not the URL exactly. Let's change `GetWebhookURL` accordingly before making it part of the API.

Reply:
This also makes me wonder if we should change the `status.url` field to `status.webhookPath` to be more appropriate.

Reply:
I'm not for breaking the API in this version. We could consider this before v1 and maybe keep both fields for 6 months?

Reply:
Doing it gradually, keeping both, sounds good to me.

Reply:
> This is a new API method and it returns the webhook path, not the URL exactly.

It returns the relative URL, without the hostname. How about `GetWebhookRelativeURL`?

Reply:
I think `status.webhookPath` and `GetWebhookPath` could make sense here - as this is only a partial URL.

Reply:
I've renamed the function to `GetWebhookPath`, please open an issue for the API changes targeting v1.

Reply:
Why not introduce `status.webhookPath` in v1beta2 as discussed above along with other changes so that later we only remove the old field? We know we will do this change and we are introducing new API here, why wait for v1?
It's the status API which is managed by the controller. It won't even be a breaking change that requires user action.

Reply:
Ok I'll introduce it in this PR and mark URL as deprecated.